modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/ParserServiceImpl.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.sql;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Supplier;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.ignite.internal.sql.engine.SqlQueryType;
+import org.apache.ignite.internal.sql.engine.util.Commons;
+
+/**
+ * An implementation of {@link ParserService} that, apart of parsing, introduces cache of parsed results.
+ */
+public class ParserServiceImpl implements ParserService {
+    private final ConcurrentMap<String, ParsedResult> queryToParsedResultCache;
+
+    /**
+     * Constructs the object.
+     *
+     * @param cacheSize The size of the cache to keep parsed results.
+     */
+    public ParserServiceImpl(int cacheSize) {
+        this.queryToParsedResultCache = Caffeine.newBuilder()
+                .maximumSize(cacheSize)
+                .<String, ParsedResult>build()
+                .asMap();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public ParsedResult parse(String query) {
+        ParsedResult result = queryToParsedResultCache.get(query);
+
+        if (result != null) {
+            return result;
+        }
+
+        StatementParseResult parsedStatement = IgniteSqlParser.parse(query, StatementParseResult.MODE);
+
+        SqlNode parsedTree = parsedStatement.statement();
+
+        SqlQueryType queryType = Commons.getQueryType(parsedTree);
+
+        assert queryType != null : parsedTree.toString();
+
+        result = new ParsedResultImpl(
+                queryType,
+                query,
+                parsedTree.toString(),
+                parsedStatement.dynamicParamsCount(),
+                () -> IgniteSqlParser.parse(query, StatementParseResult.MODE).statement()
```

Comment:
Is it a good idea to parse each non-cached query twice? :thinking:

Reply:
The very first query will be parsed twice, which looks acceptable. 
The next queries will be parsed once if either parse result will be evicted from parser cache or a plan was evicted from planner cache. 

I think we can reuse `parsedTree` for the queries, which plans shouldn't be cached, because `parsedTree` is not shared object in that case.

Reply:
> The very first query will be parsed twice.
It seems to me that this can be fixed with simple atomic bool/ref, something like
```
() -> staleTreeFlag.compareAndSet(false, true) ? parsedTree : IgniteSqlParser.parse(query, StatementParseResult.MODE).statement()
```

Reply:
> It seems to me that this can be fixed with simple atomic bool/ref, something like

Or have 2 implementations of ParsedResult.

Reply:
Folks, I agree that this looks weird at the first glance. But here is my justification of current approach:

1. we have 4 types of queries: QUERY, DML, DDL, and EXPLAIN
2. first two should pass an optimisation step, which takes from dozens to thousands milliseconds. I believe an extra millisecond won't make any difference here.
3. Both DDL and EXPLAIN  neither hot nor recurrent. Besides, DDL processing times mostly dictated by Delay Duration time, which is hundreds of milliseconds. Thus I believe an extra millisecond won't make any difference here as well.
4. An extra IF branches | ternary operators | any exceptional handling increases complexity of the code while brings (almost) no value

If you still believe it does make sense to avoid double parsing, I will be happy to improve this